File: backend/app/schemas/memory.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any

class ExceptionMemoryDTO(BaseModel):
    exception_id: str
    vendor_id: str
    exception_type: str
    root_cause: str
    resolution: str
    timestamp: str

class MemoryEventLogDTO(BaseModel):
    event_id: str
    vendor_id: str
    vendor: str
    trigger_event: str
    previous_risk: int
    updated_risk: int
    delta: int
    trend: str
    exception_type: str
    timestamp: str
    log_message: str

class VendorBehavioralProfileDTO(BaseModel):
    vendor_id: str
    vendor: str
    transactions: int
    exceptions: int
    top_issue: str
    risk_score: int
    trend: str  # "Increasing", "Decreasing", "Stable"
    avg_transaction_value: float
    duplicate_payment_count: int
    tax_mismatch_count: int
    settlement_delay_count: int
    last_updated: str
    recent_exceptions: List[ExceptionMemoryDTO] = Field(default_factory=list)
    recent_events: List[MemoryEventLogDTO] = Field(default_factory=list)

class VendorListResponseDTO(BaseModel):
    total_vendors_tracked: int
    high_risk_vendors: int
    profiles: List[VendorBehavioralProfileDTO]
    latest_events: List[MemoryEventLogDTO] = Field(default_factory=list)

class MemoryUpdateRequestDTO(BaseModel):
    vendor_id: str
    vendor_name: str
    transaction_amount: float
    has_exception: bool
    exception_type: Optional[str] = None
    root_cause: Optional[str] = None
    resolution: Optional[str] = None

class RecordExceptionRequestDTO(BaseModel):
    vendor_id: str
    vendor_name: str
    exception_type: str  # "Tax Mismatch", "Settlement Delay", "Duplicate Payment", "Unregistered Vendor"
    transaction_amount: float
    root_cause: Optional[str] = None
    resolution: Optional[str] = None

File: apps/merchant-portal/src/types/memory.ts
export interface ExceptionMemoryDTO {
  exception_id: string;
  vendor_id: string;
  exception_type: string;
  root_cause: string;
  resolution: string;
  timestamp: string;
}

export interface MemoryEventLogDTO {
  event_id: string;
  vendor_id: string;
  vendor: string;
  trigger_event: string;
  previous_risk: number;
  updated_risk: number;
  delta: number;
  trend: string;
  exception_type: string;
  timestamp: string;
  log_message: string;
}

export interface VendorBehavioralProfileDTO {
  vendor_id: string;
  vendor: string;
  transactions: number;
  exceptions: number;
  top_issue: string;
  risk_score: number;
  trend: 'Increasing' | 'Decreasing' | 'Stable' | string;
  avg_transaction_value: number;
  duplicate_payment_count: number;
  tax_mismatch_count: number;
  settlement_delay_count: number;
  last_updated: string;
  recent_exceptions: ExceptionMemoryDTO[];
  recent_events: MemoryEventLogDTO[];
}

export interface VendorListResponseDTO {
  total_vendors_tracked: number;
  high_risk_vendors: number;
  profiles: VendorBehavioralProfileDTO[];
  latest_events: MemoryEventLogDTO[];
}

export interface RecordExceptionRequestDTO {
  vendor_id: string;
  vendor_name: string;
  exception_type: string;
  transaction_amount: number;
  root_cause?: string;
  resolution?: string;
}
